include/aie/Dialect/AIE/IR/AIE.td

```diff
@@ -130,11 +130,11 @@ def AIE_TileOp: AIE_Op<"tile", [FlowEndPoint]>, Results<(outs Index:$result)> {
   }];
 
   let extraClassDeclaration = [{
-    int getNumSourceConnections(WireBundle bundle);
-    int getNumDestConnections(WireBundle bundle);
-    int colIndex() { return getCol(); }
-    int rowIndex() { return getRow(); }
-    TileID getTileID() { return std::make_pair(getCol(), getRow()); }
+    uint32_t getNumSourceConnections(WireBundle bundle);
```

Comment:
Why not match what MLIR returns? All `I8/I32/I64` inputs/outputs/attributes are the corresponding `uint`.

Reply:
I'm not entirely sure why the value of signless attributes is returned as unsigned, particularly since signless integers are treated as signed when sign extension operations are materialized; very odd.  Anyhow, yes, since unsigned integers, particularly when mixed with signed integers, can introduce a host of hard to detect and debug bugs, and the only actual disadvantage is reduced range, if there's no reason to use unsigned types it's always better to use signed. I'd argue that even if you find yourself in a situation where int32_t won't provide enough range, it's preferable to go to int64_t rather than uint32_t. As a personal rule of thumb, I reserve unsigned types for bitmasks... or at least I try, old habits die hard 😝

Reply:
Okay makes sense. I'll undo most of these then.

Reply:
ok restored all `int`s but just fyi there's now a bunch of narrowing conversion warnings: https://gist.github.com/makslevental/a300ade3aa86874344e5fb846ace4fbe

Reply:
😩 Well... that's inconvenient... let me think about this.

Reply:
If find it very strange, and I wonder whether it's a mistake or what. Look at: `mlir/include/mlir/IR/Builder.h:129`:
```c++
IntegerAttr getI32IntegerAttr(int32_t value);```

There is an inconsistency that I don't understand 🤔

Reply:
`I32` isn't signed, it's signless: https://github.com/llvm/llvm-project/blob/ca23c933bda9bdb4ef4cc8204834167f5f031401/mlir/include/mlir/IR/CommonAttrConstraints.td#L252-L253

```
def I32Attr : TypedSignlessIntegerAttrBase<
    I32, "uint32_t", "32-bit signless integer attribute">;
```

Reply:
I understand that. What I don't understand is why if you do "getInt()" or try to build a IntegerAttr, it will return/take a signed int, but when tablegen generates the default attribute getters it performs a ZExt that returns an unsigned version (which is what's tripping us here).

Reply:
> To clarify: "unsigned" is not "singless" either.

of course but there are no signless ints in C++ so between `uint` and `int`, for the concept of "signless", I indeed would expect `uint`.

Reply:
If you look pretty much everywhere else, you'll see we use "int64_t" as the canonical signless integer (See: IntegerAttr, defined in `mlir/include/mlir/IR/BuiltinAttributes.td`). It's the tablegen definition of I<X> that's not consistent with the underlying IntegerAttr class, and I'm wondering why.

Reply:
I don't think this says what you think it says?

```
    /// Return the integer value as a 64-bit int. The attribute must be a
    /// signless integer.
    // TODO: Change callers to use getValue instead.
    int64_t getInt() const;
```

that doesn't say `int` is canonical for signless, this say (to me) you can return signless as `int` (got me why). But I read the attr constraint, the thing that appears in basically every dialect op spec, as the canonical definition. And of course the fact that for every other dialect that uses `I32Attr` has `uint32_t` returend from their getters (and has been).

Reply:
I'm saying it's "canonical" because building a singless integer attribute or obtaining the value of a signless integer attribute, uses singed integers. You have to pick one or the other, because it's neither, and we seem to be picking singed everywhere **except** for the I\<X\>Attr family of constraints.

I believe the reason is because, internally, IntegerAttr is use APInt to store the value. APInt is an LLVM type, and it is meant to store an explicitly unsigned value:
```cpp
/// Class for arbitrary precision integers.
///
/// APInt is a functional replacement for common case unsigned integer type like
/// "unsigned", "unsigned long" or "uint64_t",
```
So that's where the inconsistency is coming from. If you ask me, generating a value getter for a signless attribute should, for consistency, be cast as a signed int, so it returns signlessness like the underlying attribute. Since, I'm not entirely sure of the implications or whether it's preferable to dealing with this issue for each individual case, I think we should draw expert's attention to this, I bet they've think about this. In any case, at least I know why it is this, and my suggestion would be to change it for our classes 🤷 It's late over here, but we can discuss this tomorrow?

Reply:
Just to be clear - `IntegerAttr` aren't stored as `int`s - they're stored as `llvm::APInt`s: 

```cpp
struct IntegerAttrStorage : public ::mlir::AttributeStorage {
  using KeyTy = std::tuple<::mlir::Type, APInt>;
  IntegerAttrStorage(::mlir::Type type, APInt value) : type(std::move(type)), value(std::move(value)) {}
  
  ...

  ::mlir::Type type;
  APInt value;
};

APInt IntegerAttr::getValue() const {
  return getImpl()->value;
}
```

```

Reply:
> It's late over here, but we can discuss this tomorrow?

Of course, no rush - I'm only responding because you're responding :)

> we seem to be picking singed everywhere except for the I<X>Attr family of constraints.

What I'm saying is I'm not familiar with where these other places are?

> IntegerAttr is using APInt to store the value.

Yes and thus the storage isn't `int`, it's `APInt`, which maybe has signed or unsigned semantics but it's not `int` semantics. 

> so it treats signlessness like the underlying class.

And how is that? I don't see in APInt anything that enforces "unsigned" (lots of signed methods...).

> my suggestion would be to change it for our classes

Your suggestion is to change to `SI32Attr`? I think that's a weird choice to specify a language/compiler feature just to accomodate sloppy arithmetic (e.g., people, myself included, not being able to do unsigned arithmetic in their head).

Anyway I'll do whatever you tell me to do (I'm not opinionated, just confused about what the expected behavior/result should be with all of these `static_cast`s that have to inevitably be performed).

Reply:
I've mentioned it above. The C++ class underlying an `I<X>Attr` is `IntegerAttr`. `IntegerAttr` represents all three cases: signed, unsigned, and signless. It comes with getters for all three cases: signed/unsigned/singless (`getSInt()`/`getUInt()`/`getInt()`), and for "signless" it's returning `int64_t`. Similarly, if you go to the base `Builder` class, building a signless `IntegerAttr` takes `int64_t`.

`I<X>Attr`, which is a type constraint that accepts "\<X\>-bit signless integers", when it access the value stored in the underlying `IntegerAttr` through the `APInt` it contains, it performs a zero-extension. I see how it may be the most reasonable thing to do, but in order to be consistent with the other classes (the ones that build and represent singless integers in MLIR), maybe that result should then be cast as signed.

I'm not trying to be opinionated here either, just trying to understand an inconsistency and figure out which way to solve it (if at all, I really like PEP 8's warning about consistency). If I were 100% certain about the right thing to do, I'd say we should redefine col/row as SI32Attr (with the >0 constraint it already has). Whatever the right answer to the consistency issue may be, it might still be the right thing to do, but I honestly don't know. It does feel like the kind of thing people smarter than me already spent a few hours discussing years ago 😃

Reply:
Just double checking my hunch: APInt stores using `uint` https://github.com/llvm/llvm-project/blob/1d95a071d6fc43fb413a0d3f5a9d1e52a18abab0/llvm/include/llvm/ADT/APInt.h#L1866

```cpp
  union {
    uint64_t VAL;   ///< Used to store the <= 64 bits integer value.
    uint64_t *pVal; ///< Used to store the >64 bits integer value.
  } U;
```

So the underlying relation representation is indeed unsigned (though, this of course doesn't explain why the zext is necessary 🤷‍♂️).

Reply:
As I pointed out before, `APInt` is supposed to store an _unsigned_ value (of arbitrary precision), **not** a _singless_ one. `IntegerAttr` **is** the class that stores a _signless_ integer. And when you access its value through `APInt` you need to do `zext` because if your attribute is supposed to be 8-bit signless, but you're returning it in a 64-bit "container", while doing a zero-extension may be somewhat wrong, doing a sign extension is **_definitely_** wrong 😂

Reply:
FWIW here, I implicitly decided that regardless of representation we were unlikely to have a number of rows and columns where the size really mattered (hence 'int')

Reply:
Will be handled in the follow up refactor.